### DISCUSSION TITLE
Format of custom node configuration file for local testnet

### DISCUSSION THREAD

**Discussion by Alexzander-Stone** [aptos-labs/aptos-developer-discussions#97] [Questions]
Asked by `misterblonde` on Discord (#dev-discussions):

> hey when running the local testnet for aptos using aptos node run-local-testnet  how do I specify the --config-path does anyone have a sample file I can use for my custom config? I can't find any examples anywhere. thank you. i found the validator.yaml and validator-identity.yaml but I don't know how i can set features like the lockup. 

Keywords
local testnet, node, config, format, yaml

**Comment by Alexzander-Stone** [comment]
When you run `aptos init` and `aptos node run-local-testnet --with-indexer-api` in your root directory, you'll see a new directory called `.aptos`.

A few levels below the `.aptos` directory, you'll see a file with the name `node.yaml`. This file is essentially the pre-formatted [NodeConfig](https://github.com/aptos-labs/aptos-core/blob/a270a1be68c7070e8ce452e8a202f9e9c5cda610/config/src/config/node_config.rs#L35) file that you're asking for.
> In my project structure, I found this at `.aptos/testnet/0/node.yaml`.

I recommend using this as your template for your custom config. 
For additional reference, the [NodeConfig](https://github.com/aptos-labs/aptos-core/blob/a270a1be68c7070e8ce452e8a202f9e9c5cda610/config/src/config/node_config.rs#L35) structure definition will come in use.
> E.g. I created a `config` directory, and duplicated the `node.yaml` file into it. I modified the API's address to be `"0.0.0.0:1212"`.
> When running `aptos node run-local-testnet --with-indexer-api --config-path ./node_config/node.yaml --force-restart`, the CLI returns the below console log, indicating the port was successfully changed to `1212`.
```
Node API is ready. Endpoint: http://127.0.0.1:1212/
```

**Reply by juliaaschmidt** [reply]
How in that node.yaml file do I specify the OLC length?

I would like to set these pool features:
min_stake: 0
max_stake: 100000
min_lockup_duration_secs: 0
max_lockup_duration_secs: 2592000
epoch_duration_secs: 86400
As attempted in this open bug: https://github.com/aptos-labs/aptos-core/issues/3153

**Reply by Alexzander-Stone** [reply]
> How in that node.yaml file do I specify the OLC length?
> 
> I would like to set these pool features: min_stake: 0 max_stake: 100000 min_lockup_duration_secs: 0 max_lockup_duration_secs: 2592000 epoch_duration_secs: 86400 As attempted in this open bug: [aptos-labs/aptos-core#3153](https://github.com/aptos-labs/aptos-core/issues/3153).

To achieve a similar effect as @noprom on [aptos-labs/aptos-core#3154](https://github.com/aptos-labs/aptos-core/issues/3153), you'll need to modify the `layout.yaml` that is used during [custom Genesis blob creation](https://aptos.dev/tools/aptos-cli/use-cli/cli-genesis/#generating-genesis).

Example of noprom's `layout.yaml`:
```
---
root_key: "F22409A93D1CD12D2FC92B5F8EB84CDCD24C348E32B3E7A720F3D2E288E63394"
users:
  - "jarvis"
chain_id: 40
min_stake: 0
max_stake: 100000
min_lockup_duration_secs: 0
max_lockup_duration_secs: 2592000
epoch_duration_secs: 86400
initial_lockup_timestamp: 1656615600
min_price_per_gas_unit: 1
allow_new_validators: true 
```
*For more configuration parameters, check out the [Layout struct](https://github.com/aptos-labs/aptos-core/blob/83a4430e47b1742f1e0a5ad957df156ccc68d3c4/crates/aptos-genesis/src/config.rs#L28).*

Then, you place the resulting `genesis.blob` and `waypoint.txt` within your local testnode.
* `genesis.blob` location -> `.aptos/testnet/0/genesis.blob`
* `waypoint.txt` location -> `.aptos/testnet/waypoint.txt`

**Reply by juliaaschmidt** [reply]
Thanks for your reply. I've followed your advice, changed the layout.yaml and run the steps according to: https://aptos.dev/tools/aptos-cli/use-cli/public-network/run-public-network/#generating-genesis

It currently fails on step 2  `2. Generate your ValidatorConfiguration:`. I'm using aptos client v3.1 on Mac OS the latest version on brew (https://aptos.dev/tools/aptos-cli/install-cli/install-cli-mac). However, the `--keys-dir` argument used in the instructions does not longer exist. 

Do you have any more up-to-date instructions by any chance?

**Reply by juliaaschmidt** [reply]
These instructions seem outdated. Even if I get around the keypath issue by copying the files to the current directory. Step 2 then shows success, but no genesis.blob nor bob.yaml is created. Only a `bob/owner.yaml` and a `bob/owner.yaml` and bob/operator.yaml.

**Reply by Alexzander-Stone** [reply]
You're right that the instructions are outdated, @juliaaschmidt .

For step 2, you'll need to use the `--owner-public-identity-file` parameter, rather than the mentioned `--keys-dir`. It'll look something similar to:
```console
cargo run --package aptos genesis set-validator-configuration \
    --owner-public-identity-file ./bobs/public-keys.yaml \
    --username bob \
    --validator-host bobs_host:6180 \
    --full-node-host bobs_host:6180 \
    --local-repository-dir .
```

After that, you can run the below to generate the genesis.blob and waypoint.txt:
`cargo run --package aptos -- genesis generate-genesis --local-repository-dir .`
